### PR TITLE
koord-manager: add args for common controllers

### PR DESCRIPTION
**Pull request by saintube**
### Ⅰ. Describe what this PR does

Add command options for the common controllers. We can enable/disable the known controllers via the arg `--controllers`.

### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

**Review by jasonliu747 — APPROVED** [review]
/lgtm

**Comment by FillZpp** [comment]
Should this be in feature-gate or a command-line flag?

For example, kube-controller-manager has `--feature-gates` to enable those features in controllers and `--controllers` to indicate which controllers should be started.

**Comment by saintube** [comment]
> Should this be in feature-gate or a command-line flag?
> 
> For example, kube-controller-manager has `--feature-gates` to enable those features in controllers and `--controllers` to indicate which controllers should be started.

I think both options are acceptable since these controllers have different maturities. The main reason I chose the feature gate is that the webhook plugins are maintained in this way.

**Comment by hormes** [comment]
> --controllers to indicate which controllers should be started

I think this is more accurate. The feature gate is a function switch, and the controller is more like a program. Some behaviors inside the controller should be controlled by the feature gate.

**Comment by saintube** [comment]
> > --controllers to indicate which controllers should be started
> 
> I think this is more accurate. The feature gate is a function switch, and the controller is more like a program. Some behaviors inside the controller should be controlled by the feature gate.

@FillZpp @hormes Fixed. PTAL.  
Now we use `--controllers` to control which controllers should be added to the manager.

File: cmd/koord-manager/main.go [hunk 1]
```diff
@@ -101,6 +111,9 @@ func main() {
 	flag.BoolVar(&enablePprof, "enable-pprof", true, "Enable pprof for controller manager.")
 	flag.StringVar(&pprofAddr, "pprof-addr", ":8090", "The address the pprof binds to.")
 	flag.StringVar(&syncPeriodStr, "sync-period", "", "Determines the minimum frequency at which watched resources are reconciled.")
+	pflag.StringSliceVar(&controllers, "controllers", allControllers, fmt.Sprintf("A list of controllers to enable. '*' enables all on-by-default controllers,"+
```
Comment (eahydra) [inline cmd/koord-manager/main.go:114]: We still have a small number of controllers inside, which are controlled by feature-gate. Then follow this to modify?

File: cmd/koord-manager/main.go [hunk 2]
```diff
@@ -78,10 +80,18 @@ func init() {
 	// +kubebuilder:scaffold:scheme
 }
 
+var controllers []string
+
+var allControllers = []string{
```
Comment (eahydra) [inline cmd/koord-manager/main.go:85]: Do we need to define allControllers? Adding a controller also requires changing two variables at the same time. Consider simplifying.

Comment (saintube) [reply cmd/koord-manager/main.go]: fixed. `controllers = []string{...}` is enough

File: cmd/koord-manager/main.go [hunk 3]
```diff
@@ -78,10 +80,16 @@ func init() {
 	// +kubebuilder:scaffold:scheme
 }
 
+var controllers = []string{
```
Comment (eahydra) [inline cmd/koord-manager/main.go:83]: Here we can do a little optimization
```suggestion
var controllerAddFuncs = map[string]func(manager.Manager) error{
	nodemetric.Name:   nodemetric.Add,
	noderesource.Name: noderesource.Add,
	nodeslo.Name:      nodeslo.Add,
}

var controllers = set.StringKeySet(controllerAddFuncs)
```

**Review by eahydra — COMMENTED** [review]
/lgtm

**Review by hormes — APPROVED** [review]

**Pull request merged**